pkg/drain/nodegroup.go

```diff
@@ -204,29 +224,53 @@ func (n *NodeGroupDrainer) toggleCordon(cordon bool, nodes *corev1.NodeList) {
 			logger.Debug("no need to %s node %q", cordonStatus(cordon), node.Name)
 		}
 	}
-
 }
 
-func (n *NodeGroupDrainer) evictPods(node string) (int, error) {
-	list, errs := n.evictor.GetPodsForEviction(node)
-	if len(errs) > 0 {
-		return 0, fmt.Errorf("errs: %v", errs) // TODO: improve formatting
-	}
-	if list == nil {
-		return 0, nil
-	}
-	if w := list.Warnings(); w != "" {
-		logger.Warning(w)
-	}
-	pods := list.Pods()
-	pending := len(pods)
-	for _, pod := range pods {
-		// TODO: handle API rate limiter error
-		if err := n.evictor.EvictOrDeletePod(pod); err != nil {
-			return pending, errors.Wrapf(err, "error evicting pod: %s/%s", pod.Namespace, pod.Name)
+func (n *NodeGroupDrainer) evictPods(ctx context.Context, node string) error {
+	// Loop until context times out.  We want to continually try to remove pods
+	// from the node as their eviction status changes.
+	previousReportTime := time.Now()
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("timed out (after %s) waiting for node %q to be drained", n.waitTimeout, node)
+		default:
+			list, errs := n.evictor.GetPodsForEviction(node)
+			if len(errs) > 0 {
+				return fmt.Errorf("errs: %v", errs) // TODO: improve formatting
+			}
+			if list == nil {
+				return nil
+			}
+			pods := list.Pods()
+			if len(pods) == 0 {
+				return nil
+			}
+			if w := list.Warnings(); w != "" {
+				logger.Warning(w)
+			}
+			// This log message is important but can be noisy.  It's useful to only
+			// update on a node every minute.
+			if time.Now().Sub(previousReportTime) > time.Minute*1 && len(pods) > 0 {
+				logger.Warning("%d pods are unevictable from node %s", len(pods), node)
```

Comment:
This count also includes currently terminating pods.
I think there is value in that - but maybe the wording could be improved.

Reply:
Shouldn't this happen only in case of a failure, though? Why would there be a warning if this is still working through some pods?

Reply:
`GetPodsForEviction` returns terminating pods.  This causes the for loop to keep looping allowing this message to log.

Reply:
Ah okay.